Fixes #2627: Improve accessibility for Pin Verification Screen

vrajdesai78 opened the pull request:
## Explanation
Fixes #2627: Improve accessibility for Pin Verification screen. I have removed seperate textview which is used to show error text. Now, error is set directly to PinPassPasswordInputEditText using onTextChanged extension function. 

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### Before

https://user-images.githubusercontent.com/9396084/107003150-543c2700-67b2-11eb-97f0-636b2bd4db10.mp4

### After

https://user-images.githubusercontent.com/43074241/180941747-b5dad9b0-2ea4-45c2-9003-a4dc1ac56ff9.mp4

### Passing updated testcases
![Screenshot from 2022-07-29 23-49-25](https://user-images.githubusercontent.com/43074241/181823121-7932c429-8d02-4cf5-b751-48fcabd7c15c.png)
![Screenshot from 2022-07-29 23-52-30](https://user-images.githubusercontent.com/43074241/181823129-e6c3efdc-6183-4233-86b9-b47ccb5e7d6d.png)
![Screenshot from 2022-07-29 23-55-13](https://user-images.githubusercontent.com/43074241/181823134-12b66281-ecec-4215-96b3-5ca06bc11ad2.png)

vrajdesai78 committed:
Added error message for incorrect pin

vrajdesai78 committed:
Updated error message with testcases

rt4914 reviewed (changes requested):
From Code perspective and implementation its really good.

Not sure why tests are failing, can you run End-to-End and other robolectric tests and check.

Also, maybe just merge with latest develop.

vrajdesai78 committed:
Merge branch 'develop' into Improve-accessibility-pin-verification-screen

vrajdesai78 committed:
Just commit to rerun CI Checks

vrajdesai78 commented:
Updated video based on new changes

rt4914 reviewed (changes requested):
Please remove all `DisableAccessibilityChecks ` and corresponding TODO's for #3245 

As this PR is fixing all of them too.

vrajdesai78 commented:
All robolectric tests for PinPasswordActivityTest are passing locally even without disabling Accessibility checks. 
![Screenshot from 2022-08-04 00-33-55](https://user-images.githubusercontent.com/43074241/182695463-3ad13ae3-c87b-4f6a-bbcb-a1e7c769ceb1.png)

vrajdesai78 commented:
Unable to run robolectric tests locally. I have run espresso tests and they are passing as expected. I have tried to run all the robolectric tests of PinPasswordActivityTest but it's just loading. I have kept tests running for more than 30 mins but still not even one test is getting passed or failed.

![Screenshot from 2022-08-05 14-57-09](https://user-images.githubusercontent.com/43074241/183143153-b7da6f69-3e14-40a8-9612-4185b8853895.png)

BenHenning commented:
@vrajdesai78 are you able to run any app module Robolectric tests? If so, does the PinPasswordActivityTest pass for you on Robolectric for the develop branch (i.e. without your changes in this PR)? These details may help provide some clarity into what's happening.

vrajdesai78 commented:
I have tried with develop branch. All test cases are passing in develop branch.

rt4914 reviewed (changes requested):
I have been trying to check robolectric tests for almost a dozen hours now and these are my findings.

1. `testPinPassword_withUser_forgot_inputAdminPinAndNewPinAndOldPin_wrongPinError` this test is incorrect as `pin_password_incorrect_pin` is checked incorrectly.
2. The following four tests are unable to finish on Robolectric:
<img width="590" alt="Screenshot 2022-08-11 at 2 05 48 PM" src="https://user-images.githubusercontent.com/9396084/184095858-b7434c09-0fd2-402e-8256-9c4496518b4e.png">


3. In all these 4 tests if we replace `hasErrorText(context.resources.getString(R.string.pin_password_incorrect_pin))` with `hasNoErrorText` all tests pass which means that there is definitely some issue as the tests are not visible.

4. Also, where-ever we are trying to append the text to the input-pin TextLayoutInput, we should use

```
 onView(
        allOf(
          withId(R.id.pin_password_input_pin_edit_text),
          isDescendantOfA(withId(R.id.pin_password_input_pin))
        )
      )
```
and not 
`onView(withId(R.id.pin_password_input_pin_edit_text))`


Try to work around these cases.

Maybe directly add `@Ignore` to above mentioned 4 tests and run all robolectirc tests without any change and notice they all pass. And now start debugging the remaining 4 tests one by one.

rt4914 commented:
I have made few changes to test cases (which are incorrect changes too) just to show that all tests are either passing or failing.
The biggest culprit is `testCoroutineDispatchers.runCurrent()` at some places its blocking the test completely.

## Output
<img width="1205" alt="Screenshot 2022-08-11 at 2 45 29 PM" src="https://user-images.githubusercontent.com/9396084/184102045-f5249b0b-ed36-4b47-b1a7-24bc6ce8b4e6.png">

## Corresponding Code (definitely contains some incorrect code)

https://gist.github.com/rt4914/2982b2a4c0bd13b213959902851040b0

vrajdesai78 committed:
rerun CI checks

rt4914 commented:
@BenHenning Can you have a look at my last two comments here. Tests are not finishing in certain cases.

vrajdesai78 committed:
Merge branch 'develop' into Improve-accessibility-pin-verification-screen

BenHenning commented:
@rt4914 how consistently does ``runCurrent`` block tests? Have you been able to look at all into what it's doing when blocking?

vrajdesai78 commented:
> @rt4914 how consistently does `runCurrent` block tests? Have you been able to look at all into what it's doing when blocking?

@BenHenning I have tried with multiple scenarios and found that closeSoftKeyboard() function is the culprit which blocking our tests from running. Now, the test are running (without any blockage) but they are failing as in robolectric it is not able to check error text using hasErrorText(). We have used `hasErrorText()` in other tests but in that we are checking it once submit button is clicked but here we don't have any submit button that's why test is failing.

BenHenning commented:
> > @rt4914 how consistently does `runCurrent` block tests? Have you been able to look at all into what it's doing when blocking?
> 
> @BenHenning I have tried with multiple scenarios and found that closeSoftKeyboard() function is the culprit which blocking our tests from running. Now, the test are running (without any blockage) but they are failing as in robolectric it is not able to check error text using hasErrorText(). We have used `hasErrorText()` in other tests but in that we are checking it once submit button is clicked but here we don't have any submit button that's why test is failing.

@vrajdesai78 it doesn't make sense to me that a submit button would affect anything except maybe somehow causing closeSoftKeyboard to work, or triggering a later task that can be synchronized with TestCoroutineDispatchers. What I think you need to check next is what is actually happening when the error text should be set relative to each step in the test (which will tell if it's being set correctly before it's asserted in the scenario where it fails). If it isn't being set in time, you can look into why (if it's an async thing, you should be able to use TestCoroutineDispatchers). If it is being set in time, you'll need to investigate within Espresso and Robolectric why the assertion is failing even though the state is correct.

vrajdesai78 commented:
Keep this PR Open

vrajdesai78 commented:
I have created a debugging doc for current failing robolectric test cases. You can find it [here](https://docs.google.com/document/d/13Cu3DqJ-7qP7SQsdHCs1nVWWZBP2l10_0_urW4kczGY/edit?usp=sharing) @BenHenning @rt4914 can you help me out with this

BenHenning commented on app/src/main/res/layout/pin_password_activity.xml:
```diff
@@ -106,9 +106,10 @@
+              android:layout_height="wrap_content"
               android:layout_marginTop="2dp"
               android:hint="@{viewModel.profile.isAdmin ? @string/input_pin_password_as_admin : @string/input_pin_password_as_user}"
+              app:errorMessage="@{viewModel.errorMessage}"
```
Ah, does TextInputLayout handle the accessibility case for the error field automatically?

vrajdesai78 replied:
Yes, I think so because in other cases also we are setting error message to TextInputLayout so I think that it will handle accessibility too

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -943,9 +943,9 @@ class PinPasswordActivityTest {
+      onView(withId(R.id.pin_password_input_pin)).check(
         matches(
-          isDisplayed()
+          hasErrorText(context.resources.getString(R.string.pin_password_incorrect_pin))
```
I think it may also be sensible to test the case(s) for when this error is empty, too.

vrajdesai78 replied:
Added a testcase to check that error is empty on correct pin.

BenHenning reviewed (changes requested):
Thanks @vrajdesai78. I think technically, this PR looks quite nice. Just had a couple of thoughts.

Re: the test issue, I don't have comment access on your doc so posting my suggestion here. Have you tried adding ``testCoroutineDispatchers.runCurrent()`` before the assertion to see if it's a timing issue?

vrajdesai78 commented:
> Thanks @vrajdesai78. I think technically, this PR looks quite nice. Just had a couple of thoughts.
> 
> Re: the test issue, I don't have comment access on your doc so posting my suggestion here. Have you tried adding ``testCoroutineDispatchers.runCurrent()`` before the assertion to see if it's a timing issue?

Yes, I have tested with testCouroutineDispatchers.runCurrent() but still it's giving me the same error, infact if I use it before closeSoftKeyboard() it is stopping test from executing further. So, here testCouroutineDispatchers.runCurrent() won't help. I have also tried to add submit button and tried to set error message onclick of submit button but still test is failing. I think there might be some issue with the way I am setting error message.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -224,9 +224,9 @@ class PinPasswordActivityTest {
       closeSoftKeyboard()
       onView(withId(R.id.pin_password_input_pin_edit_text)).perform(closeSoftKeyboard())
         .perform(editTextInputAction.appendText("54321"), closeSoftKeyboard())
-      onView(withText(context.getString(R.string.pin_password_incorrect_pin))).check(
```
Out of curiosity, does this old check work if put back? We should be using the text input layout version, but I'm curious.

Also, have you checked via step debugging whether the error actually is set by the time it's checking for it to be present?

Finally, have you verified the behavior on both Robolectric and Espresso? Is it failing in the same way on both?

vrajdesai78 replied:
> Out of curiosity, does this old check work if put back? We should be using the text input layout version, but I'm curious.
> 
> Also, have you checked via step debugging whether the error actually is set by the time it's checking for it to be present?
> 
> Finally, have you verified the behavior on both Robolectric and Espresso? Is it failing in the same way on both?

Tests are passing in espresso and failing in robolectric. I have tried to check whether TextInputLayout has error text or not by asserting the TextInputLayout and in that TextInputLayout.error is returning null.

BenHenning commented:
I left a follow up comment with a few suggestions to aid in investigating @vrajdesai78. I've checked your code again and looked at TextInputLayout and nothing is jumping out to me yet, so I think we need to collect more data here.

vrajdesai78 commented:
```  
@Test 
  fun testPinPassword_withAdmin_inputWrongPin_incorrectPinShows() {
    ActivityScenario.launch<PinPasswordActivity>(
      PinPasswordActivity.createPinPasswordActivityIntent(
        context = context,
        adminPin = adminPin,
        profileId = adminId
      )
    ).use {
      closeSoftKeyboard()
      onView(withId(R.id.pin_password_input_pin_edit_text)).perform(closeSoftKeyboard())
        .perform(editTextInputAction.appendText("54321"), closeSoftKeyboard())
      testCoroutineDispatchers.runCurrent()
      onView(withText(R.string.pin_password_incorrect_pin)).check(
        matches(isDisplayed())
      )
    }
  } 
```

I have tried with above test and it's passing in espresso and failing in robolectric. Current version of test i.e. checking hasErrorText is also working on espresso but failing in robolectric.

BenHenning commented:
@vrajdesai78 what do you mean by the current version of the test is failing on Robolectric? That doesn't seem to be the case given that our CI checks are passing. Am I misunderstanding? You'll also need to step debug through the assertion pipeline to figure out why it isn't matching. What's likely happening is the view doesn't have the expected state when we think it does. If that's the case, you'll need to step debug to see why it isn't being set on time, and when it actually is being set.

vrajdesai78 committed:
Merge branch 'develop' of https://github.com/vrajdesai78/oppia-android into Improve-accessibility-pin-verification-screen

vrajdesai78 commented:
> @vrajdesai78 what do you mean by the current version of the test is failing on Robolectric? That doesn't seem to be the case given that our CI checks are passing. Am I misunderstanding? You'll also need to step debug through the assertion pipeline to figure out why it isn't matching. What's likely happening is the view doesn't have the expected state when we think it does. If that's the case, you'll need to step debug to see why it isn't being set on time, and when it actually is being set.

@BenHenning I have tried to use `testCoroutineDispatchers.advanceUntilIdle()` when we enter correct pin and it opens HomeActivity, and suprisingly I found `testPinPassword_withUser_inputCorrectPin_opensHomeActivity` test is running infinite if I use `testCoroutineDispatchers.advanceUntilIdle()`. So, I am thinking that may be advaneUntilIdle is not working properly in `PinPasswordActivityTest` and we can't conclude anything from that. I have also tried to use `testCoroutineDispatchers.advanceTimeBy(50000)` but still the test is failing. 

So, now what would you suggest, how can I debug this more.

BenHenning commented:
As mentioned in chat, I think you'll need to investigate *why* advanceUntilIdle doesn't end as that will provide context on the original failure @vrajdesai78.

vrajdesai78 commented:
![image](https://user-images.githubusercontent.com/43074241/192595447-933a05af-da7d-437c-899a-57b3552defd8.png)
As @BenHenning suggested, I have tried to debug ErrorTextMatcher and found that `TextInputLayout.error` is null as shown above. I have tried to check it with different tests, and for other tests `TextInputLayout.error` is returning error properly. 

@BenHenning can you suggest me next steps

BenHenning commented:
> ![image](https://user-images.githubusercontent.com/43074241/192595447-933a05af-da7d-437c-899a-57b3552defd8.png) As @BenHenning suggested, I have tried to debug ErrorTextMatcher and found that `TextInputLayout.error` is null as shown above. I have tried to check it with different tests, and for other tests `TextInputLayout.error` is returning error properly.
> 
> @BenHenning can you suggest me next steps

@vrajdesai78 just to check, what is the value of just the error (without the 'toString' bit)? I'm curious why it's resolving to ``"null"`` rather than ``null`` (i.e. that it's a string).

This does provide us with context, though: the error isn't being defined by the time we try to check for its presence. You'll need to step through the production code & test at the same time to figure out how far into the data pipeline the test gets before it verifies the error (evidently, it's not getting all the way to setting the error).

Apologies for being so delayed on this. I will keep this assigned to me and try to investigate it later tonight my time.

vrajdesai78 commented:
@BenHenning updated [debugging doc](https://docs.google.com/document/d/13Cu3DqJ-7qP7SQsdHCs1nVWWZBP2l10_0_urW4kczGY/edit?usp=sharing). I found that pin is not entering in `pin_password_input_pin_edit_text` which is the case of failing test case.

vrajdesai78 committed:
removed disableaccessibility checks

vrajdesai78 committed:
Resolved conflicts

vrajdesai78 committed:
optimized import

rt4914 reviewed (approved):
In this I don't have anything to block on except for test cases. But relying on @BenHenning for that part of PR.

@vrajdesai78 Approving this for now to unblock you but still assign it back to me once approved by @BenHenning 
(You can merge it, I will still do the post review).

vrajdesai78 commented:
> @BenHenning updated [debugging doc](https://docs.google.com/document/d/13Cu3DqJ-7qP7SQsdHCs1nVWWZBP2l10_0_urW4kczGY/edit?usp=sharing). I found that pin is not entering in `pin_password_input_pin_edit_text` which is the case of failing test case.

Added new findings in [this](https://docs.google.com/document/d/13Cu3DqJ-7qP7SQsdHCs1nVWWZBP2l10_0_urW4kczGY/edit?usp=sharing) debugging doc.

BenHenning commented:
Leaving an update to capture some of the findings from my investigation:
- I'm finding that both runCurrent and advanceUntilIdle stall in a livelock, and I've isolated this to Robolectric's ShadowLooper.idle().
- idle() hangs after the TextInputLayout's error text is caused.
- Upon analyzing the error text, it seems to be continuously recomputing the error text icon (even though there isn't one), and is causing the error text layout to continually be remeasured.
- ShaderLooper.idle() is known to have issues with animations specifically.
- I tried to disable animations programmatically, but this is a bit tricky to do in Robolectric and didn't seem to make a difference.
- The solution listed in https://github.com/robolectric/robolectric/issues/6158 was actually reasonable (or, at least, it let us get past the problematic code), but it causes Robolectric's ActivityController to have an issue, instead (which also happens with a vanilla advanceUntilIdle() call without the delay, and without the breaking error text above).

It seems very likely this is a problem with both Robolectric and TextInputLayout (the former doesn't handle animations well, and the latter seems to be doing potentially problematic things). I still have no idea how to work around this problem as it seems like it may not actually be an issue with animations. I think one part that especially confuses me is why real-time & delayed callback time causes it to advance past the measure, but zero-time delays does not.

BenHenning commented:
I just found a really important detail: when investigating Robolectric's LooperDelegatingScheduler I found that advanceTo() is being called with increasing long time intervals. The scheduler actually is correctly running, it just reaches a point where the delay in running takes so long that it looks like a process stall. Not yet sure *why* this is happening, though.

Edit: Ah, this is because advanceUntilIdle() attempts to execute all tasks. However, if there is no limit then we'll execute ad infinitum (though TestCoroutineDispatchers' timeout should kick in at some point). This might help me root cause what's actually scheduling those tasks.

Edit 2: Welp, looks like it might be the EditText blink animation that stalls advanceUntilIdle(). I doubt this is the issue causing runCurrent() to stall when error text is set, but I want to at least find a way to disable to blink animation.

Edit 3: Confirming that disabling the blinking cursor allows advanceUntilidle() to not stall, but it's unrelated (at least partly) to runCurrent() stalling when enabling error text.

Edit 4: It seems that the setPostFrameCallbackDelay() + advanceUntilIdle() approach for 'fixing' the error text isn't actually doing what I thought it would. It somehow causes the Robolectric main looper message queue to break & stop executing things which is why Espresso ends up hanging during view assertion time.

Edit 5: Okay, so it seems like the post frame callback delay doesn't actually do anything to fix the issue (it just caused the message queue to enter a strangely invalid blocking state that, when unblocked, triggers the livelock).

BenHenning commented:
So I got the test passing by moving the TextInputLayout outside ConstraintLayout. It seems that ConstraintLayout & TextInputLayout are causing a bug in one of them (can't be sure which). This happens by ConstraintLayout re-measuring TextInputLayout several times in such a way where TextInputLayout requests a re-layout. The combination of these behaviors seems to cause a constant back-and-forth measure process that never ends in Robolectric (when observing TextInputLayout's measures, it switches between 3-4 different widths, one of which is 0 and causes the state to revert back).

I'm not yet sure what in ConstraintLayout is actually causing this, or what we can do about it.

Edit: Welp, I found a fix. I noticed that the TextInputLayout is being set with a 0dp width which I think is triggering the bad case that breaks it in Robolectric. I suspect this is just an invalid state to put the layout in, and is just causing an inefficiency in regular runtime that's not noticed (though it's possible this is still only observable in Robolectric). I'm going to investigate to see if it's okay not to use 0dp here, and possibly prohibiting it to avoid others hitting this rather difficult-to-investigate problem.

Edit 2: New theory: add_profile_activity also has a 0dp-width TextInputLayout inside a ConstraintLayout, and it also verifies TextInputLayout error text successfully. However, a major difference is that in this case we're actually nesting ConstraintLayouts with both the TextInputLayout having a 0dp (match_constraint) width, as well as its parent ConstraintLayout.

Edit 3: Confirming that changing the parent ConstraintLayout's width to not be 0dp also fixes the issue. Gotta think about how to check for this via tests/checks.

BenHenning commented:
Not stale, I'm still working on this.

BenHenning commented:
I think you can resolve this now @vrajdesai78 by changing pin_password_activity such that it doesn't have nested ConstraintLayouts. That will avoid the problem.

vrajdesai78 committed:
Merge branch 'develop' into Improve-accessibility-pin-verification-screen

vrajdesai78 committed:
Fixed failing tests

vrajdesai78 committed:
Optimized code

vrajdesai78 commented:
@BenHenning I have fixed the failing test cases as per you suggestion by setting width to match_parent instead of 0dp. Also, added a new test to check that error is not displayed when correct pin is entered. Thanks

BenHenning commented on app/src/main/res/layout/pin_password_activity.xml:
```diff
@@ -95,7 +95,7 @@
 
           <androidx.constraintlayout.widget.ConstraintLayout
-            android:layout_width="@dimen/pin_password_activity_input_with_icon_width"
+            android:layout_width="match_parent"
```
Is this change actually correct to make? A solution that doesn't require changing dimensions would be collapsing the nested ConstraintLayouts (which seems like the more serious issue here as it is both a performance issue and complex bugs like the one this PR exposed).

vrajdesai78 replied:
I have reverted back to dimen and inside dimens.xml I have used -1px which is equivalent to match parent. I have referred this [stackoverflow](https://stackoverflow.com/a/37971851/12314934) answer. I have checked UI and I haven't found any issue with the current changes.

BenHenning reviewed (commented):
Thanks @vrajdesai78. Just had one follow-up comment about the change to pin_password_activity--PTAL.

vrajdesai78 committed:
Changed dimensions

vrajdesai78 committed:
Merged to latest develop

vrajdesai78 committed:
changed back to dimen from match_parent

BenHenning commented on app/src/main/res/values-sw600dp-land/dimens.xml:
```diff
@@ -340,7 +340,7 @@
   <dimen name="pin_password_activity_hello_text_view_margin_start">0dp</dimen>
   <dimen name="pin_password_activity_hello_text_view_margin_end">0dp</dimen>
-  <dimen name="pin_password_activity_input_with_icon_width">0dp</dimen>
+  <dimen name="pin_password_activity_input_with_icon_width">-1px</dimen>
```
For this & the other files, please include a comment explaining what this means (since it's otherwise a bit of a code smell since negative width doesn't make sense).

vrajdesai78 replied:
Added a comment, thanks

BenHenning reviewed (commented):
Thanks @vrajdesai78. Just had one follow-up nit otherwise the change LGTM.

vrajdesai78 committed:
Merge branch 'develop' into Improve-accessibility-pin-verification-screen

vrajdesai78 committed:
Added comment

vrajdesai78 committed:
Merge branch 'Improve-accessibility-pin-verification-screen' of https://github.com/vrajdesai78/oppia-android into Improve-accessibility-pin-verification-screen

BenHenning reviewed (approved):
Thanks @vrajdesai78! LGTM.
